observer/probers/dns/dns_conf.go

```diff
@@ -0,0 +1,137 @@
+)
+
+// DNSConf is exported to receive YAML configuration
+type DNSConf struct {
+	Proto   string `yaml:"protocol"`
+	Server  string `yaml:"server"`
+	Recurse bool   `yaml:"recurse"`
+	QName   string `yaml:"query_name"`
+	QType   string `yaml:"query_type"`
+}
+
+// UnmarshalSettings takes YAML as bytes and unmarshals it to the to an
+// DNSConf object.
+func (c DNSConf) UnmarshalSettings(settings []byte) (probers.Configurer, error) {
+	var conf DNSConf
+	err := yaml.Unmarshal(settings, &conf)
+	if err != nil {
+		return nil, err
+	}
+	return conf, nil
+}
+
+func (c DNSConf) validateServer() error {
+	server := strings.Trim(strings.ToLower(c.Server), " ")
+	// Ensure `server` contains a port.
+	host, port, err := net.SplitHostPort(server)
+	if err != nil || port == "" {
+		return fmt.Errorf(
+			"invalid `server`, %q, could not be split: %s", c.Server, err)
+	}
+	// Ensure `server` port is valid.
+	portNum, err := strconv.Atoi(port)
+	if err != nil {
+		return fmt.Errorf(
+			"invalid `server`, %q, port must be a number", c.Server)
+	}
+	if portNum <= 0 || portNum > 65535 {
+		return fmt.Errorf(
+			"invalid `server`, %q, port number must be one in [1-65535]", c.Server)
+	}
+	// Ensure `server` is a valid FQDN or IPv4 / IPv6 address.
+	IPv6 := net.ParseIP(host).To16()
+	IPv4 := net.ParseIP(host).To4()
+	FQDN := dns.IsFqdn(dns.Fqdn(host))
+	if IPv6 == nil && IPv4 == nil && !FQDN {
+		return fmt.Errorf(
+			"invalid `server`, %q, is not an FQDN or IPv4 / IPv6 address", c.Server)
+	}
+	return nil
+}
+
+func (c DNSConf) validateProto() error {
+	validProtos := []string{"udp", "tcp"}
+	proto := strings.Trim(strings.ToLower(c.Proto), " ")
+	for _, i := range validProtos {
+		if proto == i {
+			return nil
+		}
+	}
+	return fmt.Errorf(
+		"invalid `protocol`, got: %q, expected one in: %s", c.Proto, validProtos)
+}
+
+func (c DNSConf) validateQType() error {
+	validQTypes = map[string]uint16{"A": 1, "TXT": 16, "AAAA": 28, "CAA": 257}
+	qtype := strings.Trim(strings.ToUpper(c.QType), " ")
+	q := make([]string, 0, len(validQTypes))
+	for i := range validQTypes {
+		q = append(q, i)
+		if qtype == i {
+			return nil
+		}
+	}
+	return fmt.Errorf(
+		"invalid `query_type`, got: %q, expected one in %s", c.QType, q)
+}
+
+// Validate ensures the configuration received by `DNSConf` is valid. If
+// the `DNSConf` cannot be validated, an error appropriate for end-user
+// consumption is returned.
+func (c DNSConf) Validate() error {
+	// validate `query_name`
+	if !dns.IsFqdn(dns.Fqdn(c.QName)) {
+		return fmt.Errorf("invalid `query_name`, %q is not an fqdn", c.QName)
+	}
+
+	// validate `server`
+	err := c.validateServer()
+	if err != nil {
+		return err
+	}
+
+	// validate `protocol`
+	err = c.validateProto()
+	if err != nil {
+		return err
+	}
+
+	// validate `query_type`
+	err = c.validateQType()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// MakeProber returns a `Prober` object for DNS queries.
+func (c DNSConf) MakeProber() probers.Prober {
```

Comment:
I have the same comment here that I had for the top-level observer object: right now, someone can construct a DNSProbe a) by calling `MakeProber()` directly or b) by building a `DNSProbe{}` manually. Both of these bypass the `.Validate()` method. Make this API bulletproof by un-exporting the fields of the DNSProbe and by calling Validate directly from this constructor. (And the same for the other prober types, obviously.)

Reply:
Yeah, this is super valid. In my first iteration all conf objects constructed the final object by calling `confObj.AsFinalObj()` and I backed away from that out of fear that it was a bit non-idiomatic. But when I got the Probers I didn't want to register all the `FinalObj` of each prober just to call `FinalObj.New()`. I should have stuck with my gut and kept a consistent construction pattern over trying to be more idiomatic.